File: backend/app/schemas/organization.py
"""Organization and membership schemas."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.models.enums import MemberRole, MembershipStatus, OrganizationStatus


class OrganizationResponse(BaseModel):
    """An organization the caller belongs to."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    slug: str
    status: OrganizationStatus
    created_at: datetime


class MemberUser(BaseModel):
    """The person behind a membership.

    A deliberately thin view of ``users``: a colleague's name and address are
    what a member list needs, and nothing else from that table belongs here.
    """

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    email: EmailStr
    first_name: str | None
    last_name: str | None


class MemberResponse(BaseModel):
    """One membership of the current organization."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user: MemberUser
    role: MemberRole
    status: MembershipStatus
    created_at: datetime


class MemberRoleUpdate(BaseModel):
    """Change one member's role.

    The organization is never part of this payload: it comes from the caller's
    verified membership, so a request cannot name a tenant it does not belong
    to.
    """

    role: MemberRole = Field(description="owner, admin or member.")

File: frontend/src/types/organization.ts
/**
 * Organization and membership contracts, mirroring `app/schemas/organization.py`.
 */

/** `app/models/enums.py::OrganizationStatus` */
export type OrganizationStatus = "active" | "suspended" | "archived";

/** `app/models/enums.py::MemberRole`, ordered from most to least privileged. */
export type MemberRole = "owner" | "admin" | "member";

/** `app/models/enums.py::MembershipStatus` */
export type MembershipStatus = "invited" | "active" | "suspended";

/** An organization as it appears inside another payload. */
export interface OrganizationSummary {
  id: string;
  name: string;
  slug: string;
  status: OrganizationStatus;
}

/** Response of `GET /api/v1/organizations/current`. */
export interface OrganizationResponse extends OrganizationSummary {
  /** ISO 8601 timestamp. */
  created_at: string;
}

/** The user behind a membership. Narrower than UserProfile: no status. */
export interface MemberUser {
  id: string;
  email: string;
  first_name: string | null;
  last_name: string | null;
}

/** One row of `GET /api/v1/organizations/current/members`. `id` is the membership. */
export interface MemberResponse {
  id: string;
  user: MemberUser;
  role: MemberRole;
  status: MembershipStatus;
  /** ISO 8601 timestamp. */
  created_at: string;
}

export interface MemberRoleUpdate {
  role: MemberRole;
}
